os/drivers/testcase/test_group.c

```diff
@@ -0,0 +1,250 @@
+	return ERROR;
+}
+
+int test_group_add_fined_remove(int cmd, unsigned long arg)
+{
+	int ret;
+	ret = test_init();
+	if (ret == ERROR) {
+		return ERROR;
+	}
+	child = group_allocchild();
+	if (child == NULL) {
+		dbg("group_allocchild failed.");
+		return ERROR;
+	}
+
+	child_pid = -1;
+	child->ch_flags = TCB_FLAG_TTYPE_TASK;
+	child->ch_pid = child_pid;
+	child->ch_status = 0;
+	/* Add the entry into the TCB list of children */
+	group_addchild(group, child);
+
+	/* cross check starts */
+	child_returned = group_findchild(group, child_pid);
+	if (child != child_returned) {
+		dbg("group_findchild failed.");
+		return ERROR;
+	}
+
+	child_returned = group_removechild(group, child_pid);
+	if (child != child_returned) {
+		dbg("group_removechild failed.");
+		return ERROR;
+	}
+
+	child_returned = group_findchild(group, child_pid);
+	if (child_returned != NULL) {
+		dbg("group_removechild failed.");
+		group_removechild(group, child_pid);
+		group_freechild(child);
+		return ERROR;
+	}
+	group_freechild(child);
+	return OK;
+}
+
+int test_group_alloc_free(int cmd, unsigned long arg)
+{
+	int ret;
+	struct child_status_s child_dummy;
+	ret = test_init();
+	if (ret == ERROR) {
+		return ERROR;
+	}
+	child = group_allocchild();
+	if (child == NULL) {
+		dbg("group_allocchild failed.");
+		return ERROR;
+	}
+	if (child->flink != NULL) {
+		dbg("group_allocchild failed.");
+		return ERROR;
+	}
+
+	child->flink = &child_dummy;
+	group_freechild(child);
+	if (child->flink == &child_dummy) {
+		dbg("group_freechild failed.");
+		return ERROR;
+	}
+	return OK;
+}
+
+int test_group_exit_child(int cmd, unsigned long arg)
+{
+	int ret;
+	ret = test_init();
+	if (ret == ERROR) {
+		return ERROR;
+	}
+	child_pid = kernel_thread("group", SCHED_PRIORITY_DEFAULT, TASK_STACKSIZE, group_exitchild_func, (char *const *)NULL);
+	if (child_pid < 0) {
+		dbg("task_create failed.");
+		return ERROR;
+	}
+
+	child = group_findchild(group, child_pid);
+	if (child == NULL) {
+		dbg("child is null.");
+		return ERROR;
+	}
+
+	sleep(3);
+
+	child_returned = group_exitchild(group);
+	if (child != child_returned) {
+		dbg("group_exitchild failed.");
+		return ERROR;
+	}
+
+	child_returned = group_removechild(group, child_pid);
+	if (child != child_returned) {
+		dbg("group_removechild failed.");
+		return ERROR;
+	}
+
+	group_freechild(child);
+	return OK;
+}
+
+int test_group_removechildren(int cmd, unsigned long arg)
+{
+	int ret;
+	ret = test_init();
+	if (ret == ERROR) {
+		return ERROR;
+	}
+	child = group_allocchild();
+	if (child == NULL) {
+		dbg("group_allocchild failed.");
+		return ERROR;
+	}
+
+	child_pid = -1;
+	child->ch_flags = TCB_FLAG_TTYPE_TASK;
+	child->ch_pid = child_pid;
+	child->ch_status = 0;
+	/* Add the entry into the TCB list of children */
+	group_addchild(group, child);
+
+	/* cross check starts */
+	child_returned = group_findchild(group, child_pid);
+	if (child != child_returned) {
+		dbg("group_findchild failed.");
+		return ERROR;
```

Comment:
Is it okay not to free the child which allocated from line 175, group_allocchild?

Reply:
I will add group_removechildren and group_freechild code for error case